src/plugins/intel_gpu/src/graph/impls/cm/vl_sdpa_opt.cpp

```diff
@@ -0,0 +1,199 @@
+            }
+
+            const auto& info = params.get_device_info();
+            const size_t CM_GRF_WIDTH = (info.arch <= gpu_arch::xe_hpc) ? 256 : 512;
+            const size_t q_step = static_cast<size_t>(std::floor(CM_GRF_WIDTH / 32));  // or 8 on Xe1
+            size_t wg_size = static_cast<size_t>(std::floor((max_seq_len + q_step - 1) / q_step));
+            int32_t need_wg_mapping = 0;
+            if (wg_size > 16) {
+                // # seq_len is too big to fit into a single work-group
+                // # will use fixed work-group size 16, process 16*16 (or 16*8 on xe1)
+                // # part of sequence, in this case, kernel needs to figure-out which part
+                // # it needs to handle
+                need_wg_mapping = 1;
+                wg_size = 16;
+            }
+
+            size_t wg_count;
+            if (need_wg_mapping) {
+                wg_count = 0;
+                const auto wg_seq_len = wg_size * q_step;
+                for (size_t i = 1; i < cu_seqlens.size(); i++) {
+                    auto start_idx = cu_seqlens[i - 1];
+                    auto end_idx = cu_seqlens[i];
+                    wg_count += static_cast<size_t>(std::floor((end_idx - start_idx + wg_seq_len - 1) / wg_seq_len));
+                }
+            } else {
+                wg_count = cu_seqlens.size() - 1;
+            }
+
+            auto& wgs = kd.params.workGroups;
+            wgs.global = {num_q_heads, wg_count * wg_size};
+            wgs.local = {1, wg_size};
+
+            std::vector<int32_t> scalars{need_wg_mapping};
+            kd.params.scalars.clear();
+            for (auto i : scalars) {
+                scalar_desc desc;
+                desc.t = scalar_desc::Types::INT32;
+                desc.v.s32 = static_cast<int32_t>(i);
+                kd.params.scalars.push_back(desc);
+            }
+        }};
+    }
+};
+
+class VLSDPAOptImpl : public PrimitiveImplCM {
+public:
+    DECLARE_OBJECT_TYPE_SERIALIZATION(ov::intel_gpu::cm::VLSDPAOptImpl)
+
+    Stage::Ptr vl_sdpa = make_stage<VLSDPAGenerator>();
+
+    VLSDPAOptImpl() : PrimitiveImplOCL(VLSDPAOptImplementationManager::get_type_info_static()) {}
+    VLSDPAOptImpl(const program_node& node, const RuntimeParams& params) : VLSDPAOptImpl() {
+        add_stage(vl_sdpa, params);
+    }
+
+    [[nodiscard]] std::unique_ptr<primitive_impl> clone() const override {
+        return make_deep_copy<VLSDPAOptImpl>(this);
+    }
+
+    cldnn::event::ptr execute(const std::vector<cldnn::event::ptr>& events, cldnn::primitive_inst& instance) override {
+        cldnn::stream& stream = instance.get_network().get_stream();
+        auto& vlsdpa_instance = reinterpret_cast<typed_primitive_inst<cldnn::vl_sdpa>&>(instance);
+        auto cu_seqlens = vl_sdpa_inst::get_mask_seqlens_from_memory(vlsdpa_instance.cu_seqlens_memory_ptr(), stream);
+
+        m_rt_params = std::make_unique<VLSDPARuntimeParams>(cu_seqlens);
```

Comment:
Any better way to avoid create VLSDPARuntimeParams frequently in execute? Since all runtime params are from primitive_inst, maybe we could refactor the infra by passing the instance forward to reduce overhead.
 @sshlyapn  How to you think? Not quite sure the design of the infra.

Reply:
@ceciliapeng2011, sure, let me share some background related to RuntimeParams
It's unique, non-copyable structure used in specific primitive implementations, designed to address custom requirements in dispatch data update logic
It should be created once during implementation's construction phase. After that, we can update its fields during each execute() call to pass information from the instance to get_dispatch_data_func() without exposing instance explicitly

Reply:
udpated